repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -106,20 +100,34 @@ class Session(interpreter: Interpreter)
     _statements.clear()
   }
 
+  private def changeState(newState: SessionState): Unit = {
+    synchronized {
+      _state = newState
+    }
+    stateChangedCallback(newState)
+  }
+
   private def executeCode(executionCount: Int, code: String): String = synchronized {
-    _state = SessionState.Busy()
+    changeState(SessionState.Busy())
+
+    def transitToIdle() = {
+      val executingLastStatement = executionCount == newStatementId.intValue() - 1
+      if (_statements.isEmpty || executingLastStatement) {
```

Comment:
IMHO this is a bit tricky to handle session state jitter, for example, if statement is sent to ReplDriver in every second, and each statement is finished in 1 second. Then you will still meet frequent session state change.

From my understanding, it would be better to simplify the session state, that will simplify the maintenance overhead and enhance the scalability.